### ISSUE TITLE
Additional Rcpp attributes

### ISSUE THREAD

**Issue by traversc** [RcppCore/Rcpp#1182]
I'd like to see a "signature" attribute that allows you to customize the R function signature in `RcppExports.R`.

I modified `src/attributes.cpp` [here](https://github.com/RcppCore/Rcpp/compare/master...traversc:master) to be able to do this. 

The syntax is like this:

```
// [[Rcpp::export( signature = {verbose = getOption("verbose")} )]]
void test1(bool verbose) {
  Rcout << "The verbose paramter is set to: ";
  Rcout << verbose << std::endl;
}
```

And in `RcppExports.R` it looks like this:

```
test1 <- function(verbose = getOption("verbose")) {
    invisible(.Call(`_RcppTest_test1`, verbose))
}
```

I wrote a bunch of [tests](https://github.com/traversc/RcppTest/blob/master/src/rcpp_test.cpp) to make sure it doesn't break. Here's a more complicated example using `sourceCpp`:

```
library(Rcpp)
sourceCpp(code='
#include <Rcpp.h>
using namespace Rcpp;
// [[Rcpp::export( rng = false, signature = {x = list("{A}", "B"), verbose = getOption("verbose")} )]]
int test_inline(List x, bool verbose) {
  Rcout << "The verbose paramter is set to: ";
  Rcout << verbose << std::endl;
  if(x.size() > 0) {
    CharacterVector first_element = x[0];
    Rcout << "The first element of list x is: ";
    Rcout << first_element << std::endl;
  }
  return x.size();
}')

test_inline()
# The verbose paramter is set to: 0
# The first element of list x is: "{A}"
# [1] 2
options(verbose=TRUE)
test_inline(list(1))
# The verbose paramter is set to: 1
# The first element of list x is: "1"
# [1] 1
```

The benefit of the "signature" attribute is that it lets you add arbitrary default parameters (lists, functions, environments, getOption, etc.). Yes you could write a wrapper function, but it would be nice to write it all inline. 

Any interest in a pull request?

**Comment by traversc** [comment]
Btw when you say that it could be fragile, do you mean it could break something internally, or that it would let users to shoot themselves in the foot? :P

**Comment by jjallaire** [comment]
This seems like a very nice addition to me. In terms of breakage, I know we use attributes a lot in tests but I don't recall how thoroughly the tests cover all the tributaries of attribute parsing. Generally though it seems like a revdep check would catch anything material (as if there were incorrect parameters generated anywhere presumably the package tests would fail).

**Comment by eddelbuettel** [comment]
"Fragile" may not be the best word -- I just meant that some things may be break subtly simply because this is such an important part of the usage and it extends it in interesting / powerful ways.  And the rev-deps mostly document that _already built_ packages still work, there are very few to no calls to `compileAttributes()` in the checking process.  But ... that may still be worth it.  The idea is intriguiging also because it allows for more expressiveness on the R side and the created interface.

**Comment by traversc** [comment]
Okay, here's the pull request: https://github.com/RcppCore/Rcpp/pull/1183

I did revdep checks on about 20 packages. Not comprehensive as it was taking a while, but I didn't see anything break.

**Comment by eddelbuettel** [comment]
I have a bit of a machinery ready using the [prrd](https://cloud.r-project.org/package=prrd) package I wrote so I'll give it a spin next week.

**Comment by eddelbuettel** [comment]
Forgot close this when we merged #1184 so closing now.  Happy New Year to all!
